src/metabase/util/infer_spaces.clj

```diff
@@ -4,22 +4,54 @@
   ;; TODO - The code in this namespace is very hard to understand. We should clean it up and make it readable.
   (:require [clojure.java.io :as io]
             [clojure.string :as s])
-  (:import java.lang.Math))
+  (:import java.lang.Math
+           java.util.Arrays))
 
 
 (def ^:const ^:private special-words ["checkins"])
 
-;; # Build a cost dictionary, assuming Zipf's law and cost = -math.log(probability).
-(def ^:private words (concat special-words (s/split-lines (slurp (io/resource "words-by-frequency.txt")))))
+(defn- slurp-words-by-frequency []
+  (concat special-words (s/split-lines (slurp (io/resource "words-by-frequency.txt")))))
 
 ;; wordcost = dict((k, log((i+1)*log(len(words)))) for i,k in enumerate(words))
-(def ^:private word-cost
-  (into {} (map-indexed (fn [idx word]
-                          [word (Math/log (* (inc idx) (Math/log (count words))))])
-                        words)))
+(defn- make-cost-map
+  "Creates a map keyed by the hash of the word and the cost as the value. The map is sorted by the hash value"
+  [words]
+  (let [log-count (Math/log (count words))]
+    (into (sorted-map)
+          (map-indexed (fn [idx word]
+                         [(hash word) (Math/log (* (inc idx) log-count))])
+                       words))))
 
-;; maxword = max(len(x) for x in words)
-(def ^:private max-word (apply max (map count words)))
+;; # Build arrays for a cost lookup, assuming Zipf's law and cost = -math.log(probability).
+;;
+;; This is structured as a let for efficiency reasons. It's reading in 120k strings and putting them into two
+;; correlated data structures. We want to ensure that those strings are garbage collected after we setup the
+;; structures and we don't want to read the file in twice
+(let [all-words (slurp-words-by-frequency)
+      sorted-words (make-cost-map all-words)]
+
+  (def ^:private ^"[I" word-hashes
+    "Array of word hash values, ordered by that hash value"
+    (int-array (keys sorted-words)))
+
+  (def ^:private ^"[D" word-cost
+    "Array of word cost doubles, ordered by the hash value for that word"
+    (double-array (vals sorted-words)))
```

Comment:
crazy idea. Do we need the double-precision here, or could we possibly make a this array of Floats and cut memory usage in half? I know Clojure has a `float-array` function but not sure how big of a difference that switch would make

Reply:
It would definitely save some space, the doubles as-is take up roughly 1 MB, I'd guess it'd be 0.5 MB after the change. The values in that array are a result of a `Math/log` call, which returns a double. I was worried about the loss of precision going from a double -> float and that potentially changing the behavior of the code. If we knew that the extra precision didn't matter, it would be a quick change to make.

Reply:
I think floats will be fine: the minimal difference between 2 consecutive probabilities is ~1E-5, while the sum of 10 largest differences between float and double probabilities is ~4.7E-6. Since all the algorithm does is sum the costs, I don't see how going to floats could meaningfully change the result.

Reply:
👍 thanks @sbelak I'll push up another PR that switches it over.

Reply:
If we want to get super frugal with memory we could also do primitive stemming (eg. collapse order, orders, ordering, ordered into one) and since I take it we mostly use this to make field names pretty, we can probably do with a much smaller vocabulary, but that would require a small change to not found function to make the cost be proportional to the length of not found string (this would also somewhat improve accuracy on synthetic words).